pkg/bgpv1/manager/reconcilerv2/preflight.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	// replace the old underlying instance with our recreated one
+	p.BGPInstance.Router = s.Router
+
+	// dump the existing config so all subsequent reconcilers perform their
+	// actions as if this is a new instance.
+	p.BGPInstance.Config = nil
+
+	// Clear the shadow state since any advertisements will be gone now that the instance has been recreated.
+	p.BGPInstance.Metadata = make(map[string]any)
+
+	return nil
+}
+
+// getLocalASN returns the local ASN for the given BGP instance. If the local ASN is defined in the desired config, it
+// will be returned. Currently, we do not support auto-ASN assignment, so if the local ASN is not defined in the
+// desired config, an error will be returned.
+func (r *PreflightReconciler) getLocalASN(p ReconcileParams) (int64, error) {
+	if p.DesiredConfig.LocalASN != nil {
+		return *p.DesiredConfig.LocalASN, nil
+	}
+
+	return -1, fmt.Errorf("missing ASN in desired config")
+}
+
+// getRouterID returns the router ID for the given ASN. If the router ID is defined in the desired config, it will
+// be returned. Otherwise, the router ID will be resolved from the ciliumnode annotations. If the router ID is not
+// defined in the annotations, the node IP from cilium node will be returned.
+func (r *PreflightReconciler) getRouterID(p ReconcileParams, asn int64) (string, error) {
+	if p.DesiredConfig.RouterID != nil {
+		return *p.DesiredConfig.RouterID, nil
+	}
+
+	// parse Node annotations into helper Annotation map
+	annoMap, err := agent.NewAnnotationMap(p.CiliumNode.Annotations)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse Node annotations for virtual router %v: %w", p.DesiredConfig.Name, err)
+	}
+
+	routerID, err := annoMap.ResolveRouterID(asn)
+	if err != nil {
+		if nodeIP := p.CiliumNode.GetIP(false); nodeIP == nil {
+			return "", fmt.Errorf("failed to get ciliumnode IP %v: %w", nodeIP, err)
+		} else {
+			routerID = nodeIP.String()
+		}
+	}
+
+	return routerID, nil
+}
+
+// getLocalPort returns the local port for the given ASN. If the local port is defined in the desired config, it will
+// be returned. Otherwise, the local port will be resolved from the ciliumnode annotations. If the local port is not
+// defined in the annotations, -1 will be returned.
+//
+// In gobgp, with -1 as the local port, bgp instance will start in non-listening mode.
+func (r *PreflightReconciler) getLocalPort(p ReconcileParams, asn int64) (int32, error) {
+	if p.DesiredConfig.LocalPort != nil {
+		return *p.DesiredConfig.LocalPort, nil
+	}
+
+	// parse Node annotations into helper Annotation map
+	annoMap, err := agent.NewAnnotationMap(p.CiliumNode.Annotations)
+	if err != nil {
+		return -1, fmt.Errorf("failed to parse Node annotations for virtual router %v: %w", p.DesiredConfig.Name, err)
+	}
+
+	localPort := int32(-1)
+	if attrs, ok := annoMap[asn]; ok {
+		if attrs.LocalPort != 0 {
+			localPort = attrs.LocalPort
+		}
+	}
```

Comment:
As discussed offline, annotations are still used for backwards compatibility. If user sets node level configuration via CiliumBGPNodeConfigOverride - it will take precedence over annotations.